### PR TITLE
[RzIL] Add lifter for 32-bit Arm, including thumb

### PR DESCRIPTION
**Pull request by thestr4ng3r**
# SQUASH THIS

### CODE REVIEW

File: librz/il/il_export.c [hunk 1]
```diff
@@ -244,11 +244,11 @@ static void il_opdmp_is_zero(RzILOpPure *op, RzStrBuf *sb, PJ *pj) {
 }
 
 static void il_opdmp_neg(RzILOpPure *op, RzStrBuf *sb, PJ *pj) {
-	il_op_param_1("~", op->op.neg, bv);
+	il_op_param_1("~-", op->op.neg, bv);
 }
 
 static void il_opdmp_lognot(RzILOpPure *op, RzStrBuf *sb, PJ *pj) {
-	il_op_param_1("lognot", op->op.lognot, bv);
+	il_op_param_1("~", op->op.lognot, bv);
```
Comment (XVilka) [inline librz/il/il_export.c]: Please extract these changes in a separate PR

Comment (thestr4ng3r) [reply librz/il/il_export.c]: Done: https://github.com/rizinorg/rizin/pull/2285

File: librz/analysis/arch/arm/arm_il32.c [hunk 2]
```diff
@@ -0,0 +1,2695 @@
...
+		ut32 idx = reg - ARM_REG_S0;
+		RzILOpBitVector *var = VARG(reg_var_name(ARM_REG_Q0 + idx / 4));
+		return UNSIGNED(32, idx % 4 ? SHIFTR0(var, UN(7, 32 * (idx % 4))) : var);
+	}
+	const char *var = reg_var_name(reg);
+	return var ? VARG(var) : NULL;
+}
+
+#define PC(addr, is_thumb)      (addr + (is_thumb ? 4 : 8))
+#define PCALIGN(addr, is_thumb) (PC(addr, is_thumb) & ~3)
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: Shouldn't `3` have a corresponding number suffix? `UL` or something like that?

File: librz/analysis/arch/arm/arm_il32.c [hunk 3]
```diff
@@ -0,0 +1,2695 @@
...
+ * IL to write the given capstone reg
+ */
+static RzILOpEffect *write_reg(arm_reg reg, RZ_OWN RZ_NONNULL RzILOpBitVector *v) {
+	rz_return_val_if_fail(v, NULL);
+	if (reg >= ARM_REG_D0 && reg <= ARM_REG_D31) {
+		ut32 idx = reg - ARM_REG_D0;
+		arm_reg qreg = ARM_REG_Q0 + idx / 2;
+		ut64 mask[2] = { UT64_MAX, UT64_MAX };
+		mask[idx % 2] = 0;
+		RzILOpBitVector *masked = LOGAND(read_reg(0, qreg), rz_il_op_new_bitv(rz_bv_new_from_bytes_le((const ut8 *)mask, 0, 128)));
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: What is 128 here?

File: librz/analysis/arch/arm/arm_il32.c [hunk 4]
```diff
@@ -0,0 +1,2695 @@
...
+ * Capstone: ARM_INS_BL, ARM_INS_BLX
+ * ARM: bl, blx
+ */
+static RzILOpEffect *bl(cs_insn *insn, bool is_thumb) {
+	RzILOpBitVector *tgt = ARG(0);
+	if (!tgt) {
+		return NULL;
+	}
+	return SEQ2(
+		SETG("lr", U32(((insn->address + insn->size) & ~1) | (is_thumb ? 1 : 0))),
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: Same, shouldn't it be `~1UL`?

File: librz/analysis/arch/arm/arm_il32.c [hunk 5]
```diff
@@ -0,0 +1,2695 @@
...
+	if (!ISREG(0) || !ISREG(1)) {
+		return NULL;
+	}
+	if (REGID(1) != ARM_REG_CPSR && REGID(1) != ARM_REG_SPSR && REGID(1) != ARM_REG_APSR) {
+		// only these regs supported
+		return NULL;
+	}
+	// There are more bits in ARM, but this is all we have:
+	return write_reg(REGID(0),
+		LOGOR(ITE(VARG("nf"), U32(1 << 31), U32(0)),
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c:1287]: Nice stair is right here!

File: librz/analysis/arch/arm/arm_il32.c [hunk 6]
```diff
@@ -0,0 +1,2695 @@
...
+	RzILOpEffect *eff = write_reg(REGID(0), res);
+	if (!eff) {
+		rz_il_op_pure_free(r0);
+		rz_il_op_pure_free(r1);
+		rz_il_op_pure_free(r2);
+		rz_il_op_pure_free(r3);
+		return NULL;
+	}
+	if (set_ge) {
+		eff = SEQ2(
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c:1871]: Is it possible to simplify this somehow?

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks good. Amazing job! It's a pity capstone isn't being very helpful for providing necessary opcode information. Evidently, It would greatly simplify some parts of the code.

File: librz/analysis/arch/arm/arm_il32.c (hunk 2 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il32.c]: Sure doesn't hurt at least.

File: librz/analysis/arch/arm/arm_il32.c (hunk 3 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il32.c]: Size of a q reg, but I am rewriting this anyway.

File: librz/analysis/arch/arm/arm_il32.c (hunk 5 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il32.c:1287]: You can thank clang-format

File: librz/analysis/arch/arm/arm_il32.c (hunk 6 above)
Comment (thestr4ng3r) [reply librz/analysis/arch/arm/arm_il32.c:1871]: Actually yes it can, good point.

**Review by XVilka — APPROVED** [review]
LGTM

**Review by XVilka — APPROVED** [review]

**Review by wargio — COMMENTED** [review]
lgtm. <3
